src/sentry/search/django/backend.py

```diff
@@ -136,12 +136,22 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
             )
 
         if tags:
-            matches = self._tags_to_filter(project, tags)
-            if not matches:
-                return queryset.none()
-            queryset = queryset.filter(
-                id__in=matches,
-            )
+            filter_tags = {k: v for k, v in six.iteritems(tags) if k[0] != '-'}
+            exclude_tags = {k[1:]: v for k, v in six.iteritems(tags) if k[0] == '-'}
+            if filter_tags:
+                matches = self._tags_to_filter(project, filter_tags)
+                if not matches:
+                    return queryset.none()
+                queryset = queryset.filter(
+                    id__in=matches,
+                )
+            if exclude_tags:
+                matches = self._tags_to_filter(project, exclude_tags)
```

Comment:
I think that it was a preventative measure to avoid a potentially expensive join when a tag is present on a very large number of distinct groups, but there's no rationale in the original change (getsentry/sentry#3486) so I suppose that's a @dcramer question.